index.bs

```diff
@@ -1350,6 +1352,41 @@ To <dfn for="XRView">obtain the projection matrix</dfn> for a given {{XRView}} |
 </div>
 
 
+Primary and Secondary Views {#primary-and-secondary-views}
+----------
+
+A [=view=] is a <dfn>primary view</dfn> if rendering to it is necessary for the immersive experience.
+
+A [=view=] is a <dfn>secondary view</dfn> if it is possible for content to choose to not render to it and still produce a working immersive experience. If content chooses to not render to these views, the user-agent MAY reconstruct them via reprojection.
+
+<div class="example">
+Examples of [=primary views=] include the main mono view of a handheld AR session, the main two stereo views of headworn AR/VR sessions, or all of the wall views of a CAVE session.
+
+Examples of [=secondary views=] include the [=first-person observer view=] used for video capture, or "quad views" where there are two views per eye with differing resolutions or fields of view.
+
+</div>
+
+<div class=note>
+While content should be written to assume that there may be any number of views, we expect a significant amount of content to make incorrect assumptions about the {{XRViewerPose/views}} array and thus break when presented with more than two views.
+
+Because user-agents may have the ability to use mechanisms like reprojection to render to these [=secondary views=] in lieu of the content, it is desirable to be able to distinguish between content that plans on handling these [=secondary views=] itself and content that is either oblivious to the existence of such [=secondary views=] or does not wish to deal with them.
+</div>
+
+User agents that expose [=secondary views=] MUST support a "<dfn for="secondary views">secondary-views</dfn>" [=feature descriptor=] as a hint. Content enabling this feature is expected to:
+
+ - Handle any nonzero number of [=views=] in the {{XRViewerPose/views}} array.
+ - Handle the existence of multiple [=views=] that have the same [=view/eye=].
+ - Handle the size of the {{XRViewerPose/views}} array changing from frame to frame. <span class=note>This can happen when video capture is enabled, for example</span>
+
+When "[=secondary views/secondary-views=]" is enabled, the user agent MAY surface any [=secondary views=] the device supports to the {{XRSession}}. The user agent MUST NOT use reprojection to reconstruct [=secondary views=] in such a case, but instead rely on whatever the content decides to render.
+
+Note: We recommend content use {{XRSessionInit/optionalFeatures}} to enable "[=secondary views/secondary-views=]" to ensure maximum compatibility.
+
+If [=secondary views=] have lower underlying frame rates, the {{XRSession}} MAY choose to do one or more of the following:
+
+ - Lower the overall frame rate of the application while the [=secondary views=] are active.
+ - Surface [=secondary views=] in the {{XRViewerPose/views}} array only for some of the frames. Implementations doing this SHOULD NOT have frames where the [=primary views=] are not present.
```

Comment:
Toggling on/off per-frame brings up another issue about whether web apps will need to prepare when they are first told to render to a secondary view.  If the app has any spin-up/spin-down (e.g. allocate/release an extra intermediate framebuffer), toggling on/off could aggravate that.  Even in the steady-on case, an app may need a frame or two to get its affairs in order before actually providing a frame for that view.

I believe we don't have any way to know if the app is submitting pixels for secondary views, right?  Once the UA asks the app to fill them in, it has to assume they are filled, and would end up with black pixels if not?  OpenXR is one step more explicit here, having apps decide to submit any given view at the end of the frame, which gives the app frames if needed to warm up.

Reply:
> I believe we don't have any way to know if the app is submitting pixels for secondary views, right? 

In the case of regular WebXR, yes we don't know.
For WebXR Layers, we will.

> Once the UA asks the app to fill them in, it has to assume they are filled, and would end up with black pixels if not? 

I believe so.

> OpenXR is one step more explicit here, having apps decide to submit any given view at the end of the frame, which gives the app frames if needed to warm up.

Yes, with layers `getViewSubImage` indicates that that the view should be submitted. We don't have that with plain WebXR.

Reply:
@thetuvix you are correct, we don't have a way of knowing in WebXR. Folks would end up with black pixels here. That might be okay for a few frames IMO.